src/main/scala/scalaz/stream/async/TaskDriver.scala

```diff
@@ -0,0 +1,164 @@
+package scalaz.stream.async
+
+import scalaz.stream.Process
+import scalaz.stream.Process.{Halt, Step, Emit, Await, Cont}
+import scalaz.{Trampoline, \/}
+import scalaz.\/._
+import scalaz.concurrent.{Strategy, Actor, Future, Task}
+import scalaz.stream.Cause
+import scalaz.stream.Cause.{Kill, EarlyCause}
+import scalaz.stream.Util._
+
+/**
+ * Asynchronous driver for Process, where `F` is specialized to TASK 
+ */
+object TaskDriver {
+
+  /**
+   * Runs the supplied task with possible chance to interrupt it.
+   * Task is interrupted when resulting callback is applied.
+   *
+   * Whenever task is interrupted, following is guaranteed:
+   * - if, there was asynchronous code of task in process, the `cb` will get invoked immediatelly with `Task.TaskInterrupted` exception
+   * - if the task is complete, then interruption is no-op
+   * - if the task is running `pure` code, that code is completed and `cb` is invoked with result of that code
+   *
+   * Whenever task is completed, following is guaranteed
+   * - if the interrupt was received while evaluating the task (pure code) then `cb` is called with result of the task evaluation
+   * - if the interrupt was received while evaluating the task (async code) then evaluated result is discarded
+   * - otherwise `cb` is called with result of task evaluatin
+   *
+   * @param task    Task to evaluate
+   * @param cb      cb to invoke after evalutation is done, or when task is interrupted
+   */
+  def runTaskAsyncInterruptibly[A](task: Task[A], cb: (Throwable \/ A) => Unit): () => Unit = {
+    sealed trait M
+    case object Interrupt extends M
+    case class Complete(r: Throwable \/ A) extends M
+    case class CompleteBind(r: Any, g: Any => Future[Throwable \/ A]) extends M
+
+    var open: Boolean = true
+
+    def complete(r: (Throwable \/ A)): Unit = {
+      open = false
+      cb(r)
+    }
+
+    def tryCompleteTask(current: Future[Throwable \/ A], actor: => Actor[M]): Unit = {
+      current.step match {
+        case Future.Now(r) => complete(r)
+        case Future.Async(onFinish) =>
+          onFinish((r: Throwable \/ A) => Trampoline.done(actor ! Complete(r)))
+        case Future.BindAsync(onFinish, g) =>
+          onFinish((r: Any) => Trampoline.done(actor ! CompleteBind(r, g)))
+
+      }
+    }
+
+    lazy val actor: Actor[M] = Actor[M]({
+      case Complete(r) if open => complete(r)
+      case CompleteBind(r, g) if open => tryCompleteTask(g(r), actor)
+      case Interrupt if open => complete(left(Task.TaskInterrupted))
+      case _ => //no-op
```

Comment:
@djspiewak I took the two tests from the PR you mentioned and
 - "runAsync onComplete on task never completing" passes
 - "runAsync independent onComplete exactly once on task eventually completing" fails on "inner finalizer invocation count"

I think the failure of the later is correct, hence the inner part was never consulted, because the await was interrupted before you entered in `rcv` here. I think this is perfectly consistent, as the interrupt behaves same way as if the `task` will fail in that case you as well won't consulte any inner cleanup. 

If you perhaps have any other cases can you maybe at them to this PR so we can assure they will execute as well? Thanks a lot!

I still believe that this implementation guarantees `exactly once` semantic. Can you please write test case in which this does not hold?  The onComplete here is executed for both scenarios when task completes with value and when task does not complete with value, but only once.   

I don't think so the `tryCompleteTask` has same semantics as `runAsyncInterruptibly` but maybe I am wrong here.  What the implementation does it puts the decision on the actor, and thus we guarantee here that we interrupt only asynchronous code and exit from the loop after very first asynchronous code being interrupted. In other words We won't interrupt any `pure` code, which I think is undesirable. Additionally by having actor in loop I think is a bit easier to reason about what gets executed before and after, but maybe its sort of taste than anything else. 

I am not an actor - lover, but I found them useful for these type of problem solving. 

As you mention, and as I explained above I think it is undesirable to interrupt the Task.delay or Bind code. I don't see any real use case for it, and for me this just brings another  unnecessary level of nondeterminism (e.g. why you would need to nondeterministically interrupt pure code?). You mention you see it as significant limitation, can you please explain this a bit more? If we really need it, than yeah, we can use standard `Task#runAsyncInterruptibly` perhaps.